[Feature][ClassNamespaceGuardRule] Add configurable rule for guarding of allowed namespaces for classes

dorrogeray opened the pull request:
Hi, this is a followup to https://github.com/rectorphp/rector/discussions/7110.

I have added `ClassNamespaceGuardRule` which can be used for:

**1. Restricting classes implementing an interface into one or more namespace regions**

Anything which implements `Symfony\Component\Form\FormTypeInterface` must always live in a namespace `App\Form\**`:

 - `Symfony\Component\Form\FormTypeInterface`
   - `App\Form\**`

<details><summary>Show more examples</summary>
<p>

Anything which implements `AutoMapperPlus\AutoMapperPlusBundle\AutoMapperConfiguratorInterface` must always live in a namespace `App\AutoMapper\**`:

- `AutoMapperPlus\AutoMapperPlusBundle\AutoMapperConfiguratorInterface`
    - `App\AutoMapper\**`

Anything which implements `Symfony\Component\Messenger\Handler\MessageHandlerInterface` must always live in a namespace `App\MessageHandler\**`:

- `Symfony\Component\Messenger\Handler\MessageHandlerInterface`
    - `App\MessageHandler\**`

Anything which implements `Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface` must always live in a namespace `App\DependencyInjection\**`:

- `Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface`
    - `App\DependencyInjection\**`

Anything which implements `Symfony\Component\DependencyInjection\EnvVarProcessorInterface` must always live in a namespace `App\EnvVar\**`:

- `Symfony\Component\DependencyInjection\EnvVarProcessorInterface`
    - `App\EnvVar\**`

Anything which implements `Symfony\Component\EventDispatcher\EventSubscriberInterface` must always live in a namespace `App\EventListener\**`:

- `Symfony\Component\EventDispatcher\EventSubscriberInterface`
    - `App\EventListener\**`

Anything which implements `Symfony\Component\Messenger\Middleware\MiddlewareInterface` must always live in a namespace `App\Messenger\Middleware\**`:

- `Symfony\Component\Messenger\Middleware\MiddlewareInterface`
    - `App\Messenger\Middleware\**`

Anything which implements `Symfony\Component\Messenger\Stamp\StampInterface` must always live in a namespace `App\Messenger\Stamp\**`:

- `Symfony\Component\Messenger\Stamp\StampInterface`
    - `App\Messenger\Stamp\**`

Anything which implements `Throwable` must always live in a namespace `App\Exception\**`:

- `Throwable`
    - `App\Exception\**`
    - `App\Services\**` (maybe we can tolerate exceptions in few other places as well)

</p>
</details>

**2. Restricting classes extending class into one or more namespace regions**

Anything which extends `Symfony\Component\Validator\Constraint` must always live in a namespace `App\Validator\Constraints\**`:

- `Symfony\Component\Validator\Constraint`
    - `App\Validator\Constraints\**`

<details><summary>Show more examples</summary>
<p>

Anything which extends `Doctrine\Migrations\AbstractMigration` must always live in a namespace `App\Migrations\**`:

- `Doctrine\Migrations\AbstractMigration`
    - `App\Migrations\**`

Anything which extends `Doctrine\ORM\EntityRepository` must always live in a namespace `App\Repository\**`:

- `Doctrine\ORM\EntityRepository`
    - `App\Repository\**`

Anything which extends `Doctrine\DBAL\Types` must always live in a namespace `App\Doctrine\Type\**`:

- `Doctrine\DBAL\Types`
    - `App\Doctrine\Type\**`

Anything which extends `Symfony\Component\Security\Core\Authorization\Voter\Voter` must always live in a namespace `App\Voter\**`:

- `Symfony\Component\Security\Core\Authorization\Voter\Voter`
    - `App\Voter\**`

Anything which extends `Sonata\AdminBundle\Controller\CRUDController` must always live in a namespace `App\Controller\Admin\**`:

- `Sonata\AdminBundle\Controller\CRUDController`
    - `App\Controller\Admin\**`

Anything which extends `Sonata\AdminBundle\Admin\AbstractAdmin` must always live in a namespace `App\Admin\**`:

- `Sonata\AdminBundle\Admin\AbstractAdmin`
    - `App\Admin\**`

</p>
</details>

Let me know what you think :slightly_smiling_face:

TomasVotruba commented on packages/phpstan-rules/src/Rules/ClassNamespaceGuardRule.php:
```diff
@@ -0,0 +1,168 @@
+                ]
+            ),
+        ]);
+    }
```
It seems this 2 examples have identical behavior.
Could you merge it to one for easier readability?

TomasVotruba commented on packages/phpstan-rules/src/Rules/ClassNamespaceGuardRule.php:
```diff
@@ -0,0 +1,168 @@
+        ]);
+    }
+
+    private function isClassLikeNameMatchedAgainstPattern(string $className, string $namespaceWildcardPattern): bool
```
This might be useful to reuse and cover with tests.
Could you extract it to service?

TomasVotruba commented on packages/phpstan-rules/src/Rules/ClassNamespaceGuardRule.php:
```diff
@@ -0,0 +1,168 @@
+        }
+
+        foreach ($this->guards as $type => $allowedNamespacePatterns) {
+            $classReflection = $this->reflectionProvider->getClass($classLikeName);
```
Thai hits performance in a loop. Could you move it above the foreach?

TomasVotruba commented:
Hi, thank you for the rule. I like the simple style easy to read and high quality 👍 

I've added few notes.

dorrogeray commented:
Hi, thanks for the feedback! :+1: 

I have updated the rule and resolved following issues:

- [feature] It is now possible to use the pattern to define the guarded subjects as well
  - Second configurable example demonstrates this possibility
  - I have added test cases covering this new possibility
- I have extracted ClassLikeNameMatcher and covered it with tests
- I have extracted ClassLikeNameFinder and covered it with tests
  - This one is used for reverse searching (find me ClassLikeNames matching this pattern). I have thought about ways in which this could be implemented, like using NodeVisitor, but then it would cause performance overhead always, and also would only support analysed files - the current implementation is a bit crude, but works, and causes performance overhead only when needed by the rule (and is cached per queried pattern)

dorrogeray commented:
@TomasVotruba its ready for review when you have a moment. 🙏

samsonasik commented on packages/phpstan-rules/src/Rules/ClassNamespaceGuardRule.php:
```diff
@@ -0,0 +1,209 @@
+            return false;
+        }
+
+        if ($isGuardedSubjectNamespacePattern) {
```
you  should can flip check by if (! $isGuardedSubjectNamespacePattern) and return true early, so next you can directly return $isGuardedSubjectSubclassOfGuardedPattern

TomasVotruba commented on packages/phpstan-rules/docs/rules_overview.md:
```diff
@@ -338,6 +338,98 @@ class SomeCommand extends Command
 
 <br>
 
+## ClassNamespaceGuardRule
+
+Define in which namespaces (using *, ** or ? glob-like pattern matching) can classes extending specified class or implementing specified interface exist
```
I think the rule name could explain the definition in more clear way. The "guard" is already included in the rule, as they usually forbid some behavior, so we can remove it. What about including "eclusive" and "extending" keywords?

dorrogeray replied:
How about `ClassExtendingExclusiveNamespaceRule`? Its really `extends` or `implements`, but for the simplicity sake?

TomasVotruba replied:
Sounds good :+1:

dorrogeray commented:
@samsonasik I have renamed the rule to `ClassExtendingExclusiveNamespaceRule` and simplified the conditions in the rule as suggested. I have also noticed that during last build php 8.x tests were failing, but I don't think its because of something that I have done in this branch.. :thinking:

OT: By the way, were you aware of this tool? https://github.com/qossmic/deptrac

I think that it will have some really nice synergies with this rule. I also like the way it can "collect" classes into "layers" using various rules (regex, directory, extends, implements, etc). I think that having this kind of powerful "categorization" of classes could benefit other tools to (I am thinking phpstan, rector). Basically so that you could apply certain rules only to certain subsets of your codebase. It might make sense for some more "controversial" rules. :thinking: I wonder if you think that it would be worthwhile pursuit to try to extract their categorization engine into separate package, so it could then be somehow bridged / reused for rector / phpstan?

TomasVotruba commented:
Thank you for the fast update :+1:  Let's run CI...

I'm awere of the deptract. It's just seems like 1 PHPStan rule turned into a tool (with both features in complexity). I prefer to use PHPStan, to any static analysis, as it's easier to learn by anyone who contributes here... and to be honest, I love it :)

TomasVotruba commented:
The line lenght is, see: https://github.com/symplify/symplify/runs/6224595544?check_suite_focus=true

The rest will be probably related to PHPStan 1.6 - @samsonasik could you check it?

samsonasik commented on composer.json:
```diff
@@ -22,7 +22,7 @@
         "phar-io/version": "^3.2",
         "phpstan/phpdoc-parser": "^1.2",
-        "phpstan/phpstan": "^1.5.6",
+        "phpstan/phpstan": "1.5.*",
```
this will make existing user with phpstan 1.6 already cannot update. 

The phpstan notice need to be resolved or ignored instead.

dorrogeray replied:
I see, thanks for the info. I will take another look when I have time. :+1:

dorrogeray commented:
I have reverted the changes in composer.json as @samsonasik suggested. The reason why the tests were not passing is that the `bleedingEdge` configuration used by tests disables the `NodeConnectingVisitor`, regardless of the service configuration like:

```neon
conditionalTags:
	PhpParser\NodeVisitor\NodeConnectingVisitor:
		phpstan.parser.richParserNodeVisitor: true
```

I have created a new `RuleRequiresNodeConnectingVisitorInterface` for tests, if the test implements it, following `bleedingEdge` configuration is overridden:

```neon
parameters:
	featureToggles:
		nodeConnectingVisitorCompatibility: true
```

I know that this is not a final fix to the `NodeConnectingVisitor` BC break, but that would be out of scope of this PR and frankly I don't feel confident enough to dive into such refactoring.

TomasVotruba commented:
Thank you for your work. That seems like a rabit hole issue to fix :+1: 

I'll check it locally and try to get this to merge, so we can make your work part of core and iterate from there on.
